### ISSUE TITLE
How to distinguish local discriminator and global discriminator

### ISSUE THREAD

**Issue by Gh1874** [HustQBW/Context-Encoder#1]
The question is shown as the title. It's hard to distinguish the local discriminator and global discriminator when reading the source code of new_train.py. Could you please share some explanation?
Besides, I'm still not sure about the train_label and train. Should they be the original images or masked/corrupted images? And I realised that the output of predict.py will be a single image, and so will the input. Could it be a directory with images?
And could you please share some instructions on using some functions? (e.g. new_train.py, I think I should run the whole code?)

**Comment by Coobiw** [comment]
Hello,thanks for your issue. I will talk about my own opinion in order.

1. About global or local discriminator
    Actually, this repo has not implement the global discriminator. In my algorithm, the fake input of the discriminator is:
`origin_img*(1-mask)+output_G*mask`
    **So that, the loss of discriminator just care the mask region.** That's why we call it local discriminator. If you want to add an additional global discriminator, you can probably just make the new global discriminator pay attention to the whole input img(not just the mask region, **just change the loss, input can be the same**).
    As for this, you can read this paper for more details. [Globally and Locally Consistent Image Completion](https://dl.acm.org/doi/10.1145/3072959.3073659)

2. train_label and train
    I'm not sure what you want to mean exactly. But I guss that you want to know the train and BP process. **The input image when training is an masked image**(processed by [RandomErasing](https://pytorch.org/vision/stable/generated/torchvision.transforms.RandomErasing.html?highlight=randomerase) in pytorch). After the foward, you will get an output whose shape is same as the input. **Then,  computing loss on the mask region with the original images.** 
    To sum up, the masked img is the input and the original img is the "label".

3. input of predict.py
    This is easy to extend, you can implement it and give me a pull request if it is  convenient for you! Sincerely thanks!

4. Intruction
    I'm really sorry that I have not write fuction doc when I open source this code. But actually, this code is easy to unstand. If possible, you can read Chapter.3.A and the Appendix of our paper and then you will know the whole pipeline of the algorithm.
The link of our paper is [https://opg.optica.org/ao/fulltext.cfm?uri=ao-60-26-8198&id=458598](https://opg.optica.org/ao/fulltext.cfm?uri=ao-60-26-8198&id=458598).
    And if you have another problem during this period, you can just give me an issue! I am glad to discuss it with you!

**Hope these are helpful for you!**

**Comment by Gh1874** [comment]
> Hello,thanks for your issue. I will talk about my own opinion in order.
> 
> 1. About global or local discriminator
>    Actually, this repo has not implement the global discriminator. In my algorithm, the fake input of the discriminator is:
>    `origin_img*(1-mask)+output_G*mask`
>    **So that, the loss of discriminator just care the mask region.** That's why we call it local discriminator. If you want to add an additional global discriminator, you can probably just make the new global discriminator pay attention to the whole input img(not just the mask region, **just change the loss, input can be the same**).
>    As for this, you can read this paper for more details. [Globally and Locally Consistent Image Completion](https://dl.acm.org/doi/10.1145/3072959.3073659)
> 2. train_label and train
>    I'm not sure what you want to mean exactly. But I guss that you want to know the train and BP process. **The input image when training is an masked image**(processed by [RandomErasing](https://pytorch.org/vision/stable/generated/torchvision.transforms.RandomErasing.html?highlight=randomerase) in pytorch). After the foward, you will get an output whose shape is same as the input. **Then,  computing loss on the mask region with the original images.**
>    To sum up, the masked img is the input and the original img is the "label".
> 3. input of predict.py
>    This is easy to extend, you can implement it and give me a pull request if it is  convenient for you! Sincerely thanks!
> 4. Intruction
>    I'm really sorry that I have not write fuction doc when I open source this code. But actually, this code is easy to unstand. If possible, you can read Chapter.3.A and the Appendix of our paper and then you will know the whole pipeline of the algorithm.
>    The link of our paper is https://opg.optica.org/ao/fulltext.cfm?uri=ao-60-26-8198&id=458598.
>    And if you have another problem during this period, you can just give me an issue! I am glad to discuss it with you!
> 
> **Hope these are helpful for you!**

Many thanks for replying so soon :). I'll follow your suggestions and propose another issue if I have any.

I'm still learning about the coding and the algorithm and have not decided which model I should choose. If I decide to modify based on your code, I'll try to modify the predict.py and share it with you anyway.

**Comment by Coobiw** [comment]
OK，best wishes for you!
